nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/client/ESRestClientFactory.java

```diff
@@ -68,12 +103,75 @@ protected ESClient createRestClient(ElasticSearchClientConfig config) {
+                                                      requestConfigBuilder -> requestConfigBuilder.setConnectTimeout(
+                                                              getConnectTimeoutMs(config)).setSocketTimeout(
+                                                                      getSocketTimeoutMs(config)))
+                                              .setMaxRetryTimeoutMillis(getConnectTimeoutMs(config));
+        if (StringUtils.isNotBlank(config.getOption(AUTH_USER_OPT))
+                || StringUtils.isNotBlank(config.getOption(KEYSTORE_PATH_OPT))) {
+            addClientCallback(config, builder);
+        }
+        RestClient lowLevelRestClient = builder.build();
         RestHighLevelClient client = new RestHighLevelClient(lowLevelRestClient);
         // checkConnection(client);
         return new ESRestClient(lowLevelRestClient, client);
     }
 
+    private void addClientCallback(ElasticSearchClientConfig config, RestClientBuilder builder) {
+        BasicCredentialsProvider credentialProvider = getCredentialProvider(config);
+        SSLContext sslContext = getSslContext(config);
+        builder.setHttpClientConfigCallback(httpClientBuilder -> {
+            if (sslContext != null) {
+                httpClientBuilder.setSSLContext(sslContext);
+            }
+            if (credentialProvider != null) {
+                httpClientBuilder.setDefaultCredentialsProvider(credentialProvider);
+            }
+            return httpClientBuilder;
+        });
+    }
+
+    protected BasicCredentialsProvider getCredentialProvider(ElasticSearchClientConfig config) {
+        if (StringUtils.isBlank(config.getOption(AUTH_USER_OPT))) {
+            return null;
+        }
+        String user = config.getOption(AUTH_USER_OPT);
+        String password = config.getOption(AUTH_PASSWORD_OPT);
+        BasicCredentialsProvider credentialsProvider = new BasicCredentialsProvider();
+        credentialsProvider.setCredentials(AuthScope.ANY, new UsernamePasswordCredentials(user, password));
+        return credentialsProvider;
+    }
+
+    protected SSLContext getSslContext(ElasticSearchClientConfig config) {
+        if (StringUtils.isBlank(config.getOption(KEYSTORE_PATH_OPT))) {
+            return null;
+        }
+        try {
+            Path keyStorePath = Paths.get(config.getOption(KEYSTORE_PATH_OPT));
+            String keyStorePass = config.getOption(KEYSTORE_PASSWORD_OPT);
+            KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+            try (InputStream is = Files.newInputStream(keyStorePath)) {
+                keyStore.load(is, keyStorePass.toCharArray());
+            }
+            SSLContextBuilder sslBuilder = SSLContexts.custom().loadTrustMaterial(keyStore, null);
+            return sslBuilder.build();
+        } catch (KeyStoreException | IOException | NoSuchAlgorithmException | CertificateException
+                | KeyManagementException e) {
```

Comment:
You can use `GeneralSecurityException` for most of those.

Reply:
:+1: